bindings/java/src/executor.rs

```diff
@@ -0,0 +1,152 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::cell::RefCell;
+use std::ffi::c_void;
+use std::future::Future;
+
+use jni::objects::{JClass, JObject};
+use jni::sys::jlong;
+use jni::{JNIEnv, JavaVM};
+use once_cell::sync::OnceCell;
+use tokio::task::JoinHandle;
+
+use crate::Result;
+
+static mut RUNTIME: OnceCell<Executor> = OnceCell::new();
+thread_local! {
+    static ENV: RefCell<Option<*mut jni::sys::JNIEnv>> = RefCell::new(None);
+}
+
+/// # Safety
+///
+/// This function could be only called by java vm when unload this lib.
+#[no_mangle]
+pub unsafe extern "system" fn JNI_OnUnload(_: JavaVM, _: *mut c_void) {
+    let _ = RUNTIME.take();
+}
+
+/// # Safety
+///
+/// This function could be only called when the lib is loaded and within an executor thread.
+pub(crate) unsafe fn get_current_env<'local>() -> JNIEnv<'local> {
+    let env = ENV
+        .with(|cell| *cell.borrow_mut())
+        .expect("env must be available");
+    JNIEnv::from_raw(env).expect("env must be valid")
+}
+
+pub enum Executor {
+    Tokio(tokio::runtime::Runtime),
+}
+
+impl Executor {
+    pub fn enter_with<F, R>(&self, f: F) -> R
```

Comment:
The blocking layer should be able to decouple from Tokio also.

Perhaps using executor-trait and support passing a different executor.